### PR TITLE
Provide an explicit API for HTTP proxy configuration

### LINKED ISSUE #45: Proxy support
See https://github.com/sensu-plugins/sensu-plugins-pagerduty/issues/4

### PR DESCRIPTION
**Pull request by orien**
Allows an HTTP proxy to be configured via initializer options. Great for those behind a corporate firewall.

Fixes #45.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,6 +61,25 @@ pagerduty.trigger(
...
+  proxy_host: "my.http.proxy.local",
+  proxy_port: 3128,
+  proxy_username: "<my-proxy-username>",
+  proxy_password: "<my-proxy-password>",
+)
+
+# Then proceed to trigger your incident
+# (sends the request to PagerDuty via the HTTP proxy)
+incident = pagerduty.trigger("incident description")
+```
```
Comment (orien) [inline README.md]: Describes how to configure an HTTP proxy.

File: lib/pagerduty.rb [hunk 2]
```diff
@@ -17,8 +17,20 @@ class Pagerduty
...
+  #
+  # @option options [String] :proxy_username username if authorization is
+  #   required to use the proxy.
+  #
+  # @option options [String] :proxy_password password if authorization is
+  #   required to use the proxy.
+  #
+  def initialize(service_key, options = {})
     @service_key = service_key
+    @transport = transport_from_options(options)
```
Comment (orien) [inline lib/pagerduty.rb]: We store the HTTP configuration in the `transport` instance.

File: lib/pagerduty.rb [hunk 3]
```diff
@@ -68,7 +84,11 @@ def trigger(description, options = {})
   #
   def get_incident(incident_key)
     fail ArgumentError, "incident_key is nil" if incident_key.nil?
-    PagerdutyIncident.new service_key, incident_key
+    PagerdutyIncident.new(
+      service_key,
+      incident_key,
+      transport: @transport,
```
Comment (orien) [inline lib/pagerduty.rb]: We want the configuration to be used by all HTTP calls to pager duty, so we need to pass the transport to all created `PagerdutyIncident`s. This'll mean the `acknowledge` and `resolve` method calls will use the proxy.

File: lib/pagerduty/http_transport.rb [hunk 4]
```diff
@@ -4,25 +4,31 @@
 
 class Pagerduty
   # @api private
-  module HttpTransport
+  class HttpTransport
```
Comment (orien) [inline lib/pagerduty/http_transport.rb]: Change this transport from a non-instantiated module to a class so we can store the proxy configuration as immutable state on an instance.

File: lib/pagerduty/http_transport.rb [hunk 5]
```diff
@@ -30,10 +36,17 @@ def self.http
...
 
-    def self.transported?(response)
-      response.is_a?(Net::HTTPSuccess) || response.is_a?(Net::HTTPRedirection)
+    def http_proxy
+      Net::HTTP.Proxy(
+        @options[:proxy_host],
+        @options[:proxy_port],
+        @options[:proxy_username],
+        @options[:proxy_password],
+      )
```
Comment (orien) [inline lib/pagerduty/http_transport.rb]: The `Proxy` method returns a dynamic class that can be used polymorphically as `Net::HTTP`. If the `proxy_host` is `nil` a proxy won't be used.

http://ruby-doc.org/stdlib-1.9.3/libdoc/net/http/rdoc/Net/HTTP.html#method-c-Proxy

**Comment by warrenseen** [comment]
Your pull requests are always a pleasure to read, @orien! :+1: from me.

**Comment by nemski** [comment]
:+1: very nice

**Comment by orien** [comment]
Thanks @warrenseen.

**Comment by nocache** [comment]
:+1:

**Pull request merged**